http-client/src/client.rs

```diff
@@ -23,6 +22,71 @@ impl HttpClient {
 		let transport = HttpTransportClient::new(target, config).map_err(|e| Error::TransportError(Box::new(e)))?;
 		Ok(Self { transport, request_id: AtomicU64::new(0) })
 	}
+
+	/// Perform a batch request towards the server.
+	///
+	/// Returns `Ok` if all requests were answered successfully.
+	/// Returns `Error` if any of the requests fails.
+	//
+	// TODO(niklasad1): maybe simplify generic `requests`, it's quite unreadable.
+	pub async fn batch_request<'a, T>(
+		&self,
+		requests: impl IntoIterator<Item = (impl Into<String>, impl Into<jsonrpc::Params>)>,
+	) -> Result<Vec<T>, Error>
+	where
+		T: DeserializeOwned,
+	{
+		let mut calls = Vec::new();
+		// NOTE(niklasad1): `ID` is not necessarily monotonically increasing.
+		let mut ids = Vec::new();
+
+		for (method, params) in requests.into_iter() {
+			let id = self.request_id.fetch_add(1, Ordering::SeqCst);
+			calls.push(jsonrpc::Call::MethodCall(jsonrpc::MethodCall {
+				jsonrpc: jsonrpc::Version::V2,
+				method: method.into(),
+				params: params.into(),
+				id: jsonrpc::Id::Num(id),
+			}));
+			ids.push(id);
+		}
+
+		let batch_request = jsonrpc::Request::Batch(calls);
+		let response = self
+			.transport
+			.send_request_and_wait_for_response(batch_request)
+			.await
+			.map_err(|e| Error::TransportError(Box::new(e)))?;
+
+		match response {
+			jsonrpc::Response::Single(_) => {
```

Comment:
This means that response was `deserialized` as single request object i.e, not batch (which has to contain an array of requests with a at least one request) which doesn't follow the spec.

So, we return a error for this instead of panicking.